R/metadata-template-dictionary.R

```diff
@@ -0,0 +1,315 @@
+#' template sheet and values sheet contains the possible enumerated values for
+#' those keys.
+#'
+#' @param templates A vector of metadata template synIDs.
+#' @param annotations Data frame of annotation dictionary. Required to have
+#' the columns: key, description, value, valueDescription, source.
+#' @param syn Synapse client object.
+#' @param directory The directory to download and save new versions of the
+#' metadata templates to. Defaults to `"."`.
+#' @return List of Synapse file objects, one for each metadata synID.
+#'
+#' @export
+#' @examples
+#' \dontrun{
+#' # Fake templates and annotations
+#' temps <- c("syn111111", "syn222222")
+#' annots <- data.frame(
+#'   key = c("first_name", "last_name", "last_name", "age"),
+#'   description = c(
+#'     "A person's given name",
+#'     "A person's family name",
+#'     "A person's family name",
+#'     "A person's age"
+#'   ),
+#'   value = c(NA, "Smith", "Robinson", NA),
+#'   valueDescription = c(
+#'     NA,
+#'     "From the Smith family",
+#'     "From the Robinson family",
+#'     NA
+#'   ),
+#'   source = c(NA, "Self-defined", "Self-defined", NA)
+#' )
+#'
+#' # Get Synapse client object and login
+#' synapse <- reticulate::import("synapseclient")
+#' syn <- synapse$Synapse()
+#' # This style of login only works with a Synapse config file
+#' syn$Login()
+#'
+#' update_template_dictionaries(
+#'   templates = temp,
+#'   annotations = annots,
+#'   syn = syn
+#' )
+#' }
+update_template_dictionaries <- function(templates, annotations, syn,
+                                         directory = ".") {
+  purrr::map(templates, function(x) {
+    local_file <- syn$get(x, downloadLocation = directory)
+    updated <- add_dictionary_sheets(local_file$path, annotations = annotations)
+    local_file
+  })
+}
+
+#' @title Verify Dictionary Structure
+#'
+#' @description Checks that no key has more than one description
+#' or columnType. Throws error if the dictionary is invalid.
+#'
+#' @param dictionary Dataframe with columns 'key', 'description',
+#' and 'columnType'.
+#'
+#' @export
+#' @importFrom magrittr %>%
+#' @importFrom rlang .data
+#' @examples
+#' # Valid
+#' dat <- data.frame(
+#'   key = c("first_name", "first_name"),
+#'   description = c(
+#'     "First name of individual",
+#'     "First name of individual"
+#'   ),
+#'   columnType = c("string", "string")
+#' )
+#' verify_dictionary_structure(dat)
+#'
+#' \dontrun{
+#' # Invalid -- throws error due to descriptions of first_name
+#' dat <- data.frame(
+#'   key = c("first_name", "first_name"),
+#'   description = c(
+#'     "First name of individual",
+#'     "Last name of individual"
+#'   ),
+#'   columnType = c("string", "string")
+#' )
+#' verify_dictionary_structure(dat)
+#' }
+verify_dictionary_structure <- function(dictionary) {
+  # Check that dictionary exists and has needed elements
+  if (is.null(dictionary) || is.na(dictionary)) {
```

Comment:
We're seeing an error in the example here that includes the message

```
----------- FAILURE REPORT -------------- 
   --- failure: length > 1 in coercion to logical ---
   --- srcref --- 
```

I think the `is.na(dictionary)` here is what's causing the issue. Below is what `is.na()` returns for the data frame in the example:

``` r
dat1 <- data.frame(
  key = c("first_name", "first_name"),
  description = c(
    "First name of individual",
    "Last name of individual"
  ),
  columnType = c("string", "string")
)
is.na(dat1)
#>        key description columnType
#> [1,] FALSE       FALSE      FALSE
#> [2,] FALSE       FALSE      FALSE
```

However the data frame could be all NAs and would fail this check even if it had the right columns:

```r
dat2 <- data.frame(
  key = c(NA, NA),
  description = c(NA, NA),
  columnType = c(NA, NA)
)
is.na(dat2)
#>       key description columnType
#> [1,] TRUE        TRUE       TRUE
#> [2,] TRUE        TRUE       TRUE

is.null(dat2) || is.na(dat2)
#> [1] TRUE
```

This check seems redundant since if someone explicitly calls `verify_dictionary_structure(NULL)` or `verify_dictionary_structure(NA)`, the next line would catch it since they're `NULL` and `NA` aren't data frames. So I think we could take out the first if block all together.

Reply:
Bah, you are totally right. I left this in after adding the next line since it was a better check, but didn't remove and didn't associate the error message with that part. 🤦 Leaving in extraneous things (and not "seeing" them) seems to be my big downfall. 😆 Going to have to use the old 'essay' method of reading things out loud. Haha.